### PR TITLE
Make otel scheduler sync

### PR DESCRIPTION
**Pull request by wildum**
The fix in https://github.com/grafana/alloy/pull/2027 did not address all the race conditions that could happen with the scheduler. We noticed it when getting rid of our Otel fork brought back the flakiness of a batch processor test.

With this change, the scheduling of otel components is not an async operation anymore. This simplifies the logic and ensures that the components are started before they can start consuming.

### CODE REVIEW

**Review by ptodev — COMMENTED** [review]
The [OTel contract](https://github.com/open-telemetry/opentelemetry-collector/blob/main/component/component.go#L14) states that Start must return quickly, so I suppose sync would be fine. Do we know if the Collector does this async or sync? It'd be best to just do what it does, just to be on the safe side.

Regarding the issue with the batch processor - does it happen because something is wrong on the OTel side, or because #2027 maybe doesn't work for some edge case?

**Comment by wildum** [comment]
> The [OTel contract](https://github.com/open-telemetry/opentelemetry-collector/blob/main/component/component.go#L14) states that Start must return quickly, so I suppose sync would be fine. Do we know if the Collector does this async or sync? It'd be best to just do what it does, just to be on the safe side.

It's done sync in the otel collector. The service starts everything sync (https://github.com/open-telemetry/opentelemetry-collector/blob/main/service/service.go#L230) and it's called sync here (https://github.com/open-telemetry/opentelemetry-collector/blob/main/otelcol/collector.go#L228)

> Regarding the issue with the batch processor - does it happen because something is wrong on the OTel side, or because https://github.com/grafana/alloy/pull/2027 maybe doesn't work for some edge case?

Piotr and I had a discussion about it here: https://raintank-corp.slack.com/archives/C02GSU8SHBN/p1733846257348439
TLDR is that on Update we will create a new underlying unstarted Otel component. Because we start it async, there is always a window where the lazy consumer is unpaused and the unstarted component might start consuming before being started.

**Review by ptodev — APPROVED** [review]
Do we need to add a changelog entry? If it's a problem end users could see, then mentioning the error in the changelog could indicate to them that they shouldn't see it again.

**Comment by thampiotr** [comment]
> It might be good for @thampiotr to opine here, given that this PR reverts changes in #2027.

It doesn't revert all the changes, it's reusing the approach.

**Review by thampiotr — APPROVED** [review]
Looks good, I like it more. Thanks for fixing @wildum 🚀

**Review by ptodev — DISMISSED** [review]

File: internal/component/otelcol/internal/scheduler/scheduler.go [hunk 1]
```diff
@@ -77,63 +53,28 @@ func NewWithPauseCallbacks(l log.Logger, onPause func(), onResume func()) *Sched
...
 
-	select {
-	case cs.newComponentsCh <- struct{}{}:
-		// Queued new message.
-	default:
-		// A message is already queued for refreshing running components so we
-		// don't have to do anything here.
-	}
+	level.Debug(cs.log).Log("msg", "scheduling components", "count", len(cc))
+	cs.schedComponents = cs.startComponents(ctx, h, cc...)
```
[the remark this answers is not in the document]
Comment (wildum) [reply internal/component/otelcol/internal/scheduler/scheduler.go:64]: I don't think that start should be called during Run because Run is called async and Start should be sync. This is a point that was made in this Otel PR: https://github.com/open-telemetry/opentelemetry-collector/pull/11594.

I think that starting in the New() function is fine. A lot of other components do it (see this loki.source.file for example https://github.com/grafana/alloy/blob/main/internal/component/loki/source/file/file.go#L122).

I would be open to have an extra "start" lifecycle step for all components like Otel to avoid starting anything in the New() functions.

[the remark this answers is not in the document]
Comment (ptodev) [reply internal/component/otelcol/internal/scheduler/scheduler.go:64]: > I don't think that start should be called during Run because Run is called async and Start should be sync. 

This is fine - Run can spin off start in a goroutine.

> see this loki.source.file for example

I vaguely remember a very, very old slack thread where it was pointed out that this is an antipattern. Unfortunately, I don't see an open issue on it and can't find the slack thread. In any case, I don't think it's a good example to follow. If a user calls `New` but never calls `Start`, there could be a leak of some resource. Or, another `New`/`Start` might even fail due to an old component still doing something in the background.

> I would be open to have an extra "start" lifecycle step for all components like Otel to avoid starting anything in the New() functions.

I still don't quite understand why this is necessary. Can we not just start the OTel components in Run, and make sure that they don't accept data until they've fully started?

[the remark this answers is not in the document]
Comment (wildum) [reply internal/component/otelcol/internal/scheduler/scheduler.go:64]: > I still don't quite understand why this is necessary. Can we not just start the OTel components in Run, and make sure that they don't accept data until they've fully started?

This is making things more complicated. We had race conditions that we attempted to fix and as you can see with this issue, we still have race conditions that we need to address.

The Otel lifecycle is simple: Start the components, then Run the components

If we start the components in the Run function, we don't follow the same pattern because some components will start running before other components are started. An Otel component should not start consuming data until it started (else it can panic, this is what we have with the batch processor right now).

The fix attempt that Piotr made was to use the lazyConsumer that's used as a wrapper around the otel component consumers to block the consumption with a Pausing mechanism. This wrapper starts in the "Pause" state and it solved the first race condition that we had -> this makes sure that a component is started before it consumes data on Init.

But another race condition comes in when a component is Updated. On update, the otel component is re-created (unstarted) and set to the existing lazy consumer. The scheduler is pausing the lazy consumer to block the consumption until the new otel component is started (see here for the [processor](https://github.com/grafana/alloy/blob/main/internal/component/otelcol/internal/scheduler/scheduler.go#L120)). The problem is that because the scheduling happens async, there is a window where the lazy consumer is not paused but its underlying consumers have been updated. This means that the unstarted components can consume data.
Another race condition with the current setup is that the scheduling could actually happen before the consumer update which means that data would be dispatched to components that are no longer running.
You can find more info about the races in this thread: (more info in this [thread](https://raintank-corp.slack.com/archives/C02GSU8SHBN/p1733846257348439)).

This was hard + time consuming to debug these races. I'm convinced that the best way to avoid the races and keep the logic simple is to follow the Otel pattern: first start all the components and then run them. I'm open to suggestions if you have a solution that solves the race conditions and keep the code simple enough

[the remark this answers is not in the document]
Comment (ptodev) [reply internal/component/otelcol/internal/scheduler/scheduler.go:64]: I agree that "sync scheduling" is a good idea during updates. I just don't think it's necessary during `New`. I suspect that a lot of Alloy code is written with the assumption that `New` doesn't have side effects, and I'm worried that we might break something else.

File: internal/component/otelcol/internal/scheduler/scheduler.go [hunk 2]
```diff
@@ -77,63 +53,28 @@ func NewWithPauseCallbacks(l log.Logger, onPause func(), onResume func()) *Sched
...
+	cs.stopComponents(ctx, cs.schedComponents...)
 
-	select {
-	case cs.newComponentsCh <- struct{}{}:
-		// Queued new message.
-	default:
-		// A message is already queued for refreshing running components so we
-		// don't have to do anything here.
-	}
+	level.Debug(cs.log).Log("msg", "scheduling components", "count", len(cc))
```
Comment (ptodev) [inline internal/component/otelcol/internal/scheduler/scheduler.go:63]:
```suggestion
	level.Debug(cs.log).Log("msg", "scheduling otelcol components", "count", len(cc))
```

It's quite an ambiguous log line, but if you think there is value in keeping it I'm happy to have it as a debug log.

**Review by thampiotr — COMMENTED** [review]
Couple of comments. Do we have sufficient test coverage? Cause it's a bit hard to ensure there are no bugs by just reading 😅

File: internal/component/otelcol/internal/scheduler/scheduler.go [hunk 3]
```diff
@@ -51,89 +49,96 @@ type Scheduler struct {
...
+
+	cs.schedMut.Unlock()
 
 	// Make sure we terminate all of our running components on shutdown.
 	defer func() {
-		if !firstRun { // always handle the callbacks correctly
-			cs.onPause()
-		}
-		cs.stopComponents(context.Background(), components...)
-		cs.onResume()
```
[the remark this answers is not in the document]
Comment (wildum) [reply internal/component/otelcol/internal/scheduler/scheduler.go:108]: I don't think that the components can be in a paused state at this level thanks to the mutex.

File: internal/component/otelcol/internal/scheduler/scheduler.go [hunk 4]
```diff
@@ -51,89 +49,96 @@ type Scheduler struct {
...
+
+	// 3. Change the consumers
+	// This is can only be done after stopping the pervious components and before starting the new ones.
+	updateConsumers()
+
+	// 4. Start the new components
+	level.Debug(cs.log).Log("msg", "scheduling components", "count", len(cs.schedComponents))
+	cs.schedComponents = cs.startComponents(ctx, h, cc...)
 	cs.host = h
+	//TODO: What if the trace component failed but the metrics one didn't? Should we resume all consumers?
```
[the remark this answers is not in the document]
Comment (wildum) [reply internal/component/otelcol/internal/scheduler/scheduler.go:118]: would you be fine to keep it as a TODO for now? This is also the current behavior on main and I would like this PR to be merged before the release

**Comment by wildum** [comment]
> Do we have sufficient test coverage?

We have tests at the scheduler and at the lazyconsumer level + the batch processor "Test_Update" is really good at spotting race conditions (thanks for adding all those tests when you did the change last time btw). Changes to the components themselves are really small, I think that we are ok

File: internal/component/otelcol/internal/scheduler/scheduler.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (thampiotr) [reply internal/component/otelcol/internal/scheduler/scheduler.go:118]: yep! sounds good

File: internal/component/otelcol/internal/scheduler/scheduler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (thampiotr) [reply internal/component/otelcol/internal/scheduler/scheduler.go:108]: true, this happens to be the case right now - as long as we don't ever start pausing/resuming consumers somewhere else.... I'll let you decide - adding onResume is not hard and I don't think it has downsides.

File: internal/component/otelcol/internal/scheduler/scheduler.go [hunk 5]
```diff
@@ -51,89 +49,96 @@ type Scheduler struct {
...
+// Schedule a new set of OpenTelemetry Components to run.
+// Components will only be started when the Scheduler's Run() function has been called.
 //
-// Schedule completely overrides the set of previously running components;
-// components which have been removed since the last call to Schedule will be
-// stopped.
-func (cs *Scheduler) Schedule(h otelcomponent.Host, cc ...otelcomponent.Component) {
+// Schedule() completely overrides the set of previously running components.
+// Components which have been removed since the last call to Schedule will be stopped.
+func (cs *Scheduler) Schedule(ctx context.Context, updateConsumers func(), h otelcomponent.Host, cc ...otelcomponent.Component) {
```
[the remark this answers is not in the document]
Comment (wildum) [reply internal/component/otelcol/internal/scheduler/scheduler.go:80]: I documented the updateConsumers function

It's true that the scheduler functionality is a bit extended with the fact that it calls a function that's expected to set consumers. But its primary purpose is still to schedule the otel components. I find ConsumerScheduler a bit confusing because it missing the "component" part in the name. I would prefer to keep Scheduler. I think that in the context of otel, the fact that it also expects a function to update consumers is ok

File: internal/component/otelcol/internal/scheduler/scheduler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (wildum) [reply internal/component/otelcol/internal/scheduler/scheduler.go:108]: added the resume call back with a comment

**Review by thampiotr — APPROVED** [review]
